// lingo3d-react/lib/components/World.js
import React, { useLayoutEffect, useRef } from "react";
import { rootContainer, settings } from "lingo3d";
import index from "lingo3d";
import { preventTreeShake } from "@lincode/utils";
import Setup from "./logical/Setup";
import { useMemoOnce } from "@lincode/hooks";
import scene from "lingo3d/lib/engine/scene";
import { setResolution } from "lingo3d/lib/states/useResolution";
import { setViewportSize } from "lingo3d/lib/states/useViewportSize";
preventTreeShake(index);
export const htmlContainer = document.createElement("div");
Object.assign(htmlContainer.style, {
    position: "absolute",
    left: "0px",
    top: "0px",
    width: "100%",
    height: "100%",
    pointerEvents: "none",
    userSelect: "none"
});
const World = ({ style, className, position, children, ...rest }) => {
    const divRef = useRef(null);
    rest.wasmPath && (settings.wasmPath = rest.wasmPath);
    useMemoOnce(() => {
        for (const child of [...scene.children])
            child.userData.manager && scene.remove(child);
    });
    useLayoutEffect(() => {
        const el = divRef.current;
        if (!el)
            return;
        el.appendChild(rootContainer);
        el.appendChild(htmlContainer);
        const resizeObserver = new ResizeObserver(() => {
            const res = [el.clientWidth, el.clientHeight];
            setResolution(res);
            setViewportSize(res);
        });
        resizeObserver.observe(el);
        return () => {
            resizeObserver.disconnect();
        };
    }, []);
    return (React.createElement(React.Fragment, null,
        React.createElement(Setup, { ...rest }),
        React.createElement("div", { style: {
                width: "100%", height: "100%", position: position !== null && position !== void 0 ? position : "absolute", top: 0, left: 0, display: "flex", ...style
            } },
            React.createElement("div", { style: { height: "100%" } }, children),
            React.createElement("div", { ref: divRef, style: { height: "100%", flexGrow: 1, position: "relative", zIndex: 0 } }))));
};
export default World;
