// lingo3d/lib/interface/ISimpleObjectManager.js
import { physicsDefaults, physicsSchema } from "./IPhysics";
export const simpleObjectManagerSchema = {
    ...physicsSchema,
    onClick: Function,
    onMouseDown: Function,
    onMouseUp: Function,
    onMouseOver: Function,
    onMouseOut: Function,
    onMouseMove: Function,
    onIntersect: Function,
    onIntersectOut: Function,
    name: String,
    id: String,
    intersectIds: Array,
    width: Number,
    height: Number,
    depth: Number,
    scaleX: Number,
    scaleY: Number,
    scaleZ: Number,
    scale: Number,
    rotationX: Number,
    rotationY: Number,
    rotationZ: Number,
    rotation: Number,
    bloom: Boolean,
    reflection: Boolean,
    outline: Boolean,
    visible: Boolean,
    innerVisible: Boolean,
    frustumCulled: Boolean,
    metalnessFactor: Number,
    roughnessFactor: Number,
    environmentFactor: Number,
    toon: Boolean,
    pbr: Boolean
};
export const simpleObjectManagerDefaults = {
    ...physicsDefaults,
    name: "",
    width: 100,
    height: 100,
    depth: 100,
    scaleX: 1,
    scaleY: 1,
    scaleZ: 1,
    scale: 1,
    rotationX: 0,
    rotationY: 0,
    rotationZ: 0,
    rotation: 0,
    bloom: false,
    reflection: false,
    outline: false,
    visible: true,
    innerVisible: true,
    frustumCulled: true,
    metalnessFactor: 0,
    roughnessFactor: 1,
    environmentFactor: 1,
    toon: false,
    pbr: false
};
