// lingo3d/lib/display/core/Loaded.js
import { Resolvable } from "@lincode/promiselikes";
import { Group, Mesh } from "three";
import { boxGeometry } from "../primitives/Cube";
import { wireframeMaterial } from "../utils/reusables";
import ObjectManager from "./ObjectManager";
import { addOutline, deleteOutline } from "../../engine/renderLoop/effectComposer/outlinePass";
import { addBloom, deleteBloom } from "../../engine/renderLoop/effectComposer/selectiveBloomPass/renderSelectiveBloom";
import { addSSR, deleteSSR } from "../../engine/renderLoop/effectComposer/ssrPass";
export default class Loaded extends ObjectManager {
    constructor() {
        super(new Mesh(boxGeometry, wireframeMaterial));
        this.loadedGroup = new Group();
        this.loadedResolvable = new Resolvable();
        this.outerObject3d.add(this.loadedGroup);
    }
    get src() {
        return this._src;
    }
    set src(src) {
        if (this._src === src)
            return;
        this._src = src;
        if (!src)
            return;
        if (this.loadedResolvable.done) {
            this.loadedResolvable = new Resolvable();
            this.loadedGroup.clear();
        }
        this.load(src).then(loaded => {
            if (this.done || this._src !== src)
                return;
            this.object3d.visible = !!this._boxVisible;
            this.resolveLoaded(loaded);
        });
    }
    get onLoad() {
        return this._onLoad;
    }
    set onLoad(cb) {
        var _a;
        this._onLoad = cb;
        (_a = this.loadedHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        cb && (this.loadedHandle = this.loadedResolvable.then(cb));
    }
    get width() {
        return super.width;
    }
    set width(val) {
        super.width = val;
        this.widthSet = true;
    }
    get height() {
        return super.height;
    }
    set height(val) {
        super.height = val;
        this.heightSet = true;
    }
    get depth() {
        return super.depth;
    }
    set depth(val) {
        super.depth = val;
        this.depthSet = true;
    }
    get innerRotationX() {
        return super.innerRotationX;
    }
    set innerRotationX(val) {
        super.innerRotationX = val;
        this.loadedGroup.rotation.x = this.object3d.rotation.x;
    }
    get innerRotationY() {
        return super.innerRotationY;
    }
    set innerRotationY(val) {
        super.innerRotationY = val;
        this.loadedGroup.rotation.y = this.object3d.rotation.y;
    }
    get innerRotationZ() {
        return super.innerRotationZ;
    }
    set innerRotationZ(val) {
        super.innerRotationZ = val;
        this.loadedGroup.rotation.z = this.object3d.rotation.z;
    }
    get innerX() {
        return super.innerX;
    }
    set innerX(val) {
        super.innerX = val;
        this.loadedGroup.position.x = this.object3d.position.x;
    }
    get innerY() {
        return super.innerY;
    }
    set innerY(val) {
        super.innerY = val;
        this.loadedGroup.position.y = this.object3d.position.y;
    }
    get innerZ() {
        return super.innerZ;
    }
    set innerZ(val) {
        super.innerZ = val;
        this.loadedGroup.position.z = this.object3d.position.z;
    }
    get innerVisible() {
        return this.loadedGroup.visible;
    }
    set innerVisible(val) {
        this.loadedGroup.visible = val;
    }
    get frustumCulled() {
        return this.outerObject3d.frustumCulled;
    }
    set frustumCulled(val) {
        if (this.outerObject3d.frustumCulled === val)
            return;
        this.outerObject3d.frustumCulled = val;
        this.loadedResolvable.then(() => super.frustumCulled = val);
    }
    get physics() {
        var _a;
        return (_a = this._physics) !== null && _a !== void 0 ? _a : false;
    }
    set physics(val) {
        var _a;
        if (this._physics === val)
            return;
        this._physics = val;
        (_a = this.physicsHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        const handle = this.physicsHandle = this.cancellable();
        this.loadedResolvable.then(() => this.initPhysics(val, handle));
    }
    get boxVisible() {
        var _a;
        return (_a = this._boxVisible) !== null && _a !== void 0 ? _a : this.object3d.visible;
    }
    set boxVisible(val) {
        this._boxVisible = val;
        this.object3d.visible = val;
    }
    get outline() {
        return !!this._outline;
    }
    set outline(val) {
        var _a;
        if (this._outline === val)
            return;
        this._outline = val;
        (_a = this._outlineHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._outlineHandle = this.loadedResolvable.then(loaded => {
            val ? addOutline(loaded) : deleteOutline(loaded);
        });
    }
    get bloom() {
        return !!this._bloom;
    }
    set bloom(val) {
        var _a;
        if (this._bloom === val)
            return;
        this._bloom = val;
        (_a = this._bloomHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._bloomHandle = this.loadedResolvable.then(loaded => {
            val ? addBloom(loaded) : deleteBloom(loaded);
        });
    }
    get reflection() {
        return !!this._reflection;
    }
    set reflection(val) {
        var _a;
        if (this._reflection === val)
            return;
        this._reflection = val;
        (_a = this._reflectionHandle) === null || _a === void 0 ? void 0 : _a.cancel();
        this._reflectionHandle = this.loadedResolvable.then(loaded => {
            val ? addSSR(loaded) : deleteSSR(loaded);
        });
    }
    addToRaycastSet(set, handle) {
        handle.watch(this.loadedResolvable.then(loaded => {
            if (!this.managerSet) {
                this.managerSet = true;
                loaded.traverse(child => { var _a; var _b; return (_a = (_b = child.userData).manager) !== null && _a !== void 0 ? _a : (_b.manager = this); });
            }
            set.add(loaded);
            handle.then(() => set.delete(loaded));
        }));
    }
}
