// lingo3d/lib/api/core/PositionedItem.js
import getCenter from "../../display/utils/getCenter";
import { vector3_ } from "../../display/utils/reusables";
import { vec2Point } from "../../display/utils/vec2Point";
import { scaleUp, scaleDown } from "../../engine/constants";
import EventLoopItem from "./EventLoopItem";
export default class PositionedItem extends EventLoopItem {
    get x() {
        return this.outerObject3d.position.x * scaleUp;
    }
    set x(val) {
        this.outerObject3d.position.x = val * scaleDown;
    }
    get y() {
        return this.outerObject3d.position.y * scaleUp;
    }
    set y(val) {
        this.outerObject3d.position.y = val * scaleDown;
    }
    get z() {
        return this.outerObject3d.position.z * scaleUp;
    }
    set z(val) {
        this.outerObject3d.position.z = val * scaleDown;
    }
    getWorldPosition() {
        var _a;
        return vec2Point(((_a = this.object3d) !== null && _a !== void 0 ? _a : this.outerObject3d).getWorldPosition(vector3_));
    }
    getCenter() {
        var _a;
        return vec2Point(getCenter(((_a = this.object3d) !== null && _a !== void 0 ? _a : this.outerObject3d)));
    }
}
